### PR TITLE
Table: Support display of multiple sub tables

### PR DESCRIPTION
**Pull request by joey-grafana**
**What is this feature?**

This PR provides functionality to add nested sub tables within our table component. 

**Why do we need this feature?**

Adds support for multiple sub-tables by adding `config: { nested: true }` to the field that contains the nested data in your dataframe.

This nested fields values can contain one or more dataframes. Each of these dataframes will be rendered as one unique sub table.

For each dataframe and index in the nested field, the dataframe will be rendered as one or more sub tables below the main dataframe row at that index.

**Who is this feature for?**

Users of the table component.

**Which issue(s) does this PR fix?**:

This PR also fixes a number of issues that seemed to pre-exist in the table component with regards to the use of sub tables;
- Pagination in dashboards would not work correctly with tables that had sub tables as each page after the first was not using the correct index to select which sub table row should be sized and opened when the expander icon is clicked.
- Selecting the 'Show table footer' option in a dashboard would cause a `max call stack size exceeded error`.
- Changing the tables cell height in dashboards would render with empty space in the sub tables.

**Special notes for your reviewer:**

We have a use case in Tempo for using multiple sub tables and I have added code to this PR to demonstrate how we can use multiple sub tables.

Storybook

Screenshot 2023-07-20 at 12 55 07

Tempo

Screenshot 2023-07-20 at 13 05 10

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

# Deprecation notice

Starting with 10.2, `parentRowIndex` is deprecated. It will be removed in a future release. From 10.2, sub-tables are supported by adding `FieldType.nestedFrames` to the field that contains the nested data in your dataframe.

### CODE REVIEW

**Comment by joey-grafana** [comment]
Hey @torkelo I was speaking with Andre who worked on sub tables a while back about some ideas and improvements. We have a use case for multiple sub tables in mind (for a new tracing API from the Tempo squad) and so I drafted this PR. 

I've also updated storybook/tests to see what the data looks like.

Please let me know what you think

File: packages/grafana-data/src/types/dataFrame.ts
Comment (torkelo) [inline packages/grafana-data/src/types/dataFrame.ts:101]: I think field(s) that holds the nested frame values makes more sense and would support multiple levels of nesting. 

We already have support for data frame value fields. Just need to mark them as holding nested frames so table panel shows them when expanding a row (instead of as columns)

**Review by torkelo — COMMENTED** [review]
This definitely looks better!! 

Not 100% sure about the nested property, should it be on the standard field config or table specific (custom) ?

File: packages/grafana-data/src/types/dataFrame.ts
Comment (joey-grafana) [reply packages/grafana-data/src/types/dataFrame.ts]: If I'm understanding this correctly, then I think the idea is to group all the nested fields into one field, where each value in that single field can be an array of data frames, where each data frame is one of the nested tables for the row at that values (array of data frames) index. I've pushed a commit to address this.

**Comment by joey-grafana** [comment]
> Not 100% sure about the nested property, should it be on the standard field config or table specific (custom) ?

Yeah, I think it makes more sense to be in custom as this is specific to the table component. I've pushed a commit for this.

File: packages/grafana-data/src/types/dataFrame.ts
Comment (torkelo) [reply packages/grafana-data/src/types/dataFrame.ts]: We can have multiple fields with nested frames, I think that is fine. 

Otherwise each field value would be an array of data frames with could work but not sure how messy that is. But might be easier than dealing with multiple fields

Comment (torkelo) [reply packages/grafana-data/src/types/dataFrame.ts]: > This approach uses a column/field with a DataFrame type to hold nested data

Yes, this is what I meant.

Comment (joey-grafana) [reply packages/grafana-data/src/types/dataFrame.ts]: Ah ok so multiple fields with their own nested data is the approach I went with first in this PR, but felt it was a cleaner to have one field with all the nested data and pushed updates for that. I've left it as is for the moment - we can easily compare what changed ([de0fb7e](https://github.com/grafana/grafana/pull/71953/commits/de0fb7ee8dd6dff8f4b58899f7cbb8e331b7e717)), but happy to change it back to multiple fields if necessary.

File: packages/grafana-data/src/field/fieldOverrides.ts
Comment (torkelo) [inline packages/grafana-data/src/field/fieldOverrides.ts:220]: data, and frameIndex look incorect here (not the nested frame index, and data should be nestedFrames I think)

File: packages/grafana-data/src/field/fieldOverrides.ts
Comment (torkelo) [inline packages/grafana-data/src/field/fieldOverrides.ts:236]: hm... this is not good, if we need to apply field config (and link supplier) to nested frames I think we need another type check than using custom.nested, that was only for Table panel to know about. 

This needs a new FieldType, we already have a "frame" field type. Maybe we can add a FieldType.nestedFrames?

**Review by torkelo — COMMENTED** [review]
I wish there was a way to make this more dynamic and extensible, this many sub tables scenario feels very specific to this particular use case. Another scenario might be we want to expand a row and show a chart/graph (instead of a nested table). 

And we are missing a very big factor in making this dynamic / load on demand, so the data does not need to be queried before showing the table.  

But the dynamic / load on demand part looks very hard, esp when dealing with later refreshes and time range changes and preserving state (of the expanded rows, and re-triggering the expanded row dynamic data with the new time range)

**Comment by aocenas** [comment]
@torkelo 
> I wish there was a way to make this more dynamic and extensible, this many sub tables scenario feels very specific to this particular use case. Another scenario might be we want to expand a row and show a chart/graph (instead of a nested table).

I think this could be nicely solved by the custom cell renderer. I think the issues is that table itself should not know how to render anything else than table, but grafana could inject a custom cell renderer that would check the dataFrame, pick an appropriate visualization and use that to render the sub row as something else than table. I think though this makes sense as an injectable behaviour rather than some intrinsic functionality of the table as that would make table dependant on lots of other code.

File: packages/grafana-data/src/field/fieldOverrides.ts
Comment (joey-grafana) [reply packages/grafana-data/src/field/fieldOverrides.ts]: ahh, thank you

**Comment by torkelo** [comment]
@aocenas that is a really clever idea. I think the only thing missing is some mechanism for the Table to know the height of the custom nested field component (which can could be rendering as sub table or graph)

File: public/app/features/explore/Table/TableContainer.tsx
Comment (ifrost) [inline public/app/features/explore/Table/TableContainer.tsx:77]: Hopefully, after https://github.com/grafana/grafana/pull/71811, this should not be needed any more.

Comment (joey-grafana) [reply public/app/features/explore/Table/TableContainer.tsx]: Hey @ifrost, I've merged with main but am seeing unexpected behaviour. Clicking on a subtable link refreshes explore and opens a new page and `traceIdHidden` but I don't see the variables being replaced.

Screenshot 2023-08-01 at 09 09 48

Screenshot 2023-08-01 at 09 00 41

Comment (ifrost) [reply public/app/features/explore/Table/TableContainer.tsx]: I can have a look. Are all changes pushed to the branch? What's the easiest way to set this up to see sub tables?

Comment (joey-grafana) [reply public/app/features/explore/Table/TableContainer.tsx]: All changes pushed now. You can run `make devenv sources=tempo` and run a query like in the following image to see the sub tables. Thanks!

Screenshot 2023-08-01 at 09 59 55

File: packages/grafana-data/src/field/fieldOverrides.ts
Comment (ifrost) [inline packages/grafana-data/src/field/fieldOverrides.ts:233]:
```suggestion
                options.timeZone,
                options.dataLinkPostProcessor
```

File: public/app/features/explore/Table/TableContainer.tsx
Comment (ifrost) [reply public/app/features/explore/Table/TableContainer.tsx]: Oh, it's because getLinksSupplier requires the same post-processor to be passed to make this work in context of Explore:

Apply changes in https://github.com/grafana/grafana/pull/71953/files#r1280456828 and it should work now!

```typescript
              valueField.getLinks = getLinksSupplier(
                nestedFrames[nfIndex],
                valueField,
                valueField.state!.scopedVars,
                context.replaceVariables,
                options.timeZone,
                options.dataLinkPostProcessor
              );
```

Comment (joey-grafana) [reply public/app/features/explore/Table/TableContainer.tsx]: Ahh of course! That did it, thanks Piotr.

**Review by adrapereira — COMMENTED** [review]
Couple of comments but we're almost ready to merge!

File: packages/grafana-ui/src/components/Table/Table.tsx
Comment (joey-grafana) [reply packages/grafana-ui/src/components/Table/Table.tsx]: Beautiful, thank you

**Review by adrapereira — APPROVED** [review]
LGTM! Thank you for the changes 🙏

File: packages/grafana-ui/src/components/Table/Table.mdx
Comment (aocenas) [inline packages/grafana-ui/src/components/Table/Table.mdx:12]: From the code it seems like it always has to be an array even if it's a single frame. Can we change it to allow just a single DataFrame?

**Comment by aocenas** [comment]
Probably for separate PR but I would prefer some more distinction in the UI for the subtables. In the storybook example it looks much better imho because the background color difference in the subtable is much more pronounced (not sure why maybe because it uses overall different page background color?).

In the Tempo case it just really blends in with the rest of the table, especially with the columns in the subtable having the same widths so I am sort of mentally struggling sometimes to differentiate the subtable from the main table rows.

**Comment by ifrost** [comment]
This is a breaking change, right? The component will no longer support `parentRowIndex`. Should it be deprecated first with backwards compatibility before dropping old behavior? Or at least announced as a breaking change (with breaking change notice in the changelog)?

File: packages/grafana-ui/src/components/Table/Table.mdx
Comment (baldm0mma) [inline packages/grafana-ui/src/components/Table/Table.mdx:13]: Are these being removed/deprecated?

File: packages/grafana-ui/src/components/Table/Table.tsx
Comment (baldm0mma) [inline packages/grafana-ui/src/components/Table/Table.tsx:230]: Good styling choice!

File: packages/grafana-ui/src/components/Table/hooks.ts
Comment (baldm0mma) [inline packages/grafana-ui/src/components/Table/hooks.ts:53]:
```suggestion
        const lowestExpandedIndex = parseInt(expandedIndexes[0], 10);
```

File: packages/grafana-ui/src/components/Table/hooks.ts
Comment (baldm0mma) [inline packages/grafana-ui/src/components/Table/hooks.ts:54]:
```suggestion
        if (!isNaN(lowestExpandedIndex)) {
```

File: packages/grafana-ui/src/components/Table/Table.mdx
Comment (joey-grafana) [reply packages/grafana-ui/src/components/Table/Table.mdx:13]: Good catch. `parentRowIndex ` was added for the Trace results table and I'm not sure if it's used by anyone else. Can't be sure however. I wonder is it sufficient to add a breaking change notice to the PR & change log for this or should we add a deprecation notice and migrate? @ifrost

**Review by baldm0mma — APPROVED** [review]
Just a few nits, but over all it LGTM from a Table-owner perspective, but I'd love to get a second pair of BI Squad eyes on it also, so I'll post another reminder in our channel. :)

**Review by codeincarnate — APPROVED** [review]
Looks really good overall! Tried to find some things to nitpick but I think @baldm0mma and others have that covered for me 😆

File: packages/grafana-ui/src/components/Table/Table.mdx
Comment (ifrost) [reply packages/grafana-ui/src/components/Table/Table.mdx:13]: This is a `grafana-ui` component. Please check with the code owners (@grafana/grafana-bi-squad) on what's the deprecation policy. Might be worth checking with how we deprecate APIs in other components in `grafana-ui` with @grafana/grafana-frontend-platform.

Comment (joey-grafana) [reply packages/grafana-ui/src/components/Table/Table.mdx:13]: I added migration logic in https://github.com/grafana/grafana/pull/71953/commits/6ecbf9de3497e03da5deaa769ceee56e4976713b. The logic exists in the TablePanels adjacent `migrations.ts` file. The migration logic is used in `TableContainer.tsx` & `TablePanel.tsx`. Basically if the supplied data frames contain `meta.custom.parentRowIndex` they will be converted to the new nested format with `FieldType.nestedFrames` instead.

I also added a deprecation notice to `Table.mdx`. Not sure if this is the correct location or is enough to express a deprecation? It does not mention when the removal of the old `parentRowIndex` will take place.

Hopefully someone from the Bi Squad / Frontend Platform can offer help/guidance on how to deprecate something like this.

Comment (baldm0mma) [reply packages/grafana-ui/src/components/Table/Table.mdx:13]: I think this and a change log note should cover it.

Comment (joey-grafana) [reply packages/grafana-ui/src/components/Table/Table.mdx:13]: AFAIK, the change log is auto generated if you add the `Add to changelog` label to the PR - so in that case it just uses the PRs title.

**Comment by joey-grafana** [comment]
@torkelo @ifrost just want to be sure here, are we ok to merge?

File: packages/grafana-ui/src/components/Table/Table.tsx
Comment (torkelo) [inline packages/grafana-ui/src/components/Table/Table.tsx:228]: what is this meta.custom.noHeader property?

**Review by torkelo — COMMENTED** [review]
@joey-grafana beyond the styling issue I think this looks good, and in the right direction. I am trying to refactor this PR a bit to make the concept of nested data a bit more generic and support not just sub tables but other visualizations. 

Not sure yet how that will turn out, but will open that as a separate PR in the next few hours or tomorrow. Depending on how that goes we probably can merge this and continue that work in that PR

File: packages/grafana-ui/src/components/Table/Table.tsx
Comment (joey-grafana) [reply packages/grafana-ui/src/components/Table/Table.tsx:228]: It allows a user to hide a tables header, should they not want to show it in the nested fields for example. It's a remnant of the older sub tables work which I thought may be helpful.

**Comment by joey-grafana** [comment]
> @joey-grafana beyond the styling issue I think this looks good, and in the right direction. I am trying to refactor this PR a bit to make the concept of nested data a bit more generic and support not just sub tables but other visualizations.
> 
> Not sure yet how that will turn out, but will open that as a separate PR in the next few hours or tomorrow. Depending on how that goes we probably can merge this and continue that work in that PR

Sounds great. I appreciate that, thank you. I will hold off on merging this for today/tomorrow until we see how that goes. Also, cheers for aligning the expander icon!

**Comment by torkelo** [comment]
@joey-grafana I pushed some refactorings to this PR that should not change anything, just moves the rendering of the sub tables to a component named ExpandedRow and simplifies some logic a bit

**Comment by joey-grafana** [comment]
> @joey-grafana I pushed some refactorings to this PR that should not change anything, just moves the rendering of the sub tables to a component named ExpandedRow and simplifies some logic a bit

`ExpandedRow` component makes this much cleaner :D Works great. Might be missing something but we lost the emphasize for the table background color when in dark mode. Not required of course, I simply felt like it added clarity to the design

**Comment by torkelo** [comment]
@joey-grafana 

> Might be missing something but we lost the emphasize for the table background color when in dark mode 

This was deliberate, we should not have any different background for the expanded row. 

We should copy design from interactive table (with a line along the left side): 

Screenshot from 2023-08-09 14-06-11

We should also disable the hover background for expanded rows

**Comment by torkelo** [comment]
@joey-grafana nice!!

**Review by torkelo — APPROVED** [review]
I think we can merge this now, there are things that could need further work. 

* like adding support for other custom ways to visualize the expanded data.  (could easily be added via nested field config cell component) 
* Figuring out how to load nested data async 

But the biggest tricky bit is how to deal with the field config for the nested frames, in a world where they could be visualized differenly

File: public/app/plugins/panel/table/TablePanel.tsx
Comment (ifrost) [inline public/app/plugins/panel/table/TablePanel.tsx:19]: Could migration happen inside Table component? I'm thinking about a use case if someone has an App plugin and uses Table component directly from grafana-ui, not TablePanel. In such case migration won't happen and the only way would be to write it manually (app plugins won't be able to import migration functions).

**Review by ifrost — APPROVED** [review]

File: public/app/plugins/panel/table/TablePanel.tsx
Comment (joey-grafana) [reply public/app/plugins/panel/table/TablePanel.tsx:19]: Would need to move the migration logic to the Table component itself for the reasons you suggested. Feels kind of messy as migration logic for panels seems to be in the panel folder itself which seems to be the norm. I'm thinking that a subset of app plugins use the table component and a subset of those would use it with sub table. Even at that, the changes required are actually quite minimal for a user to migrate themselves (which they will eventually have to do anyway). This may essentially be a non issue, but can't be certain.

**Pull request merged**
